src/System.IO.Pipelines/src/System/IO/Pipelines/PipeCompletion.cs

```diff
@@ -0,0 +1,118 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.ExceptionServices;
+
+namespace System.IO.Pipelines
+{
+    [DebuggerDisplay("IsCompleted: {" + nameof(IsCompleted) + "}")]
+    internal struct PipeCompletion
+    {
+        private static readonly ArrayPool<PipeCompletionCallback> CompletionCallbackPool = ArrayPool<PipeCompletionCallback>.Shared;
+
+        private const int InitialCallbacksSize = 1;
+        private static readonly Exception _completedNoException = new Exception();
+
+        private Exception _exception;
+
+        private PipeCompletionCallback[] _callbacks;
+        private int _callbackCount;
+
+        public bool IsCompleted => _exception != null;
+
+        public PipeCompletionCallbacks TryComplete(Exception exception = null)
+        {
+            if (_exception == null)
```

Comment:
Do all Exception objects here come from a trusted source? You're calling into the Equals() method which on Exception is overridable.

Reply:
Should it be `ReferenceEquals(_exception, null)`?

Reply:
Either that or `if ((object)_exception == null)`

Reply:
On second though, forget what I said - this one isn't a problem. I was thinking of a similar situation on `Type` where we overload "==" to call "Equals."